tests/h/services/search_index/_queue_test.py

```diff
@@ -0,0 +1,364 @@
+import datetime
+import logging
+from unittest import mock
+
+import pytest
+from h_matchers import Any
+
+from h.models import Job
+from h.search.client import Client
+from h.search.index import BatchIndexer
+from h.services.search_index._queue import Queue
+
+
+class TestAddSyncAnnotationJob:
+    def test_it(self, db_session, queue):
+        queue.add("test_annotation_id", "test_tag", scheduled_at=one_week_from_now)
+
+        assert db_session.query(Job).all() == [
+            Any.instance_of(Job).with_attrs(
+                dict(
+                    enqueued_at=Any.instance_of(datetime.datetime),
+                    scheduled_at=one_week_from_now,
+                    tag="test_tag",
+                    kwargs={"annotation_id": "test_annotation_id"},
+                )
+            ),
+        ]
+
+    def test_scheduled_at_defaults_to_now(self, db_session, queue):
+        queue.add("test_annotation_id", "test_tag")
+
+        assert db_session.query(Job).all() == [
+            Any.instance_of(Job).with_attrs(
+                dict(
+                    enqueued_at=Any(),
+                    scheduled_at=Any.instance_of(datetime.datetime),
+                    tag=Any(),
+                    kwargs=Any(),
+                )
+            ),
+        ]
+
+
+class TestSyncAnnotations:
+    def test_it_does_nothing_if_the_queue_is_empty(self, batch_indexer, caplog, queue):
+        queue.sync()
+
+        batch_indexer.index.assert_not_called()
+        assert caplog.records == []
+
+    def test_it_ignores_jobs_that_arent_scheduled_yet(
+        self, annotation_ids, batch_indexer, caplog, queue
+    ):
+        queue.add_all(
+            annotation_ids,
+            tag="test",
+            scheduled_at=one_week_from_now,
+        )
+
+        queue.sync()
+
+        batch_indexer.index.assert_not_called()
+        assert caplog.records == []
+
+    def test_it_ignores_jobs_beyond_limit(self, annotation_ids, batch_indexer, queue):
+        queue.add_all(
+            annotation_ids,
+            tag="test",
+        )
+
+        queue.sync()
+
+        for annotation_id in annotation_ids[limit:]:
+            assert annotation_id not in batch_indexer.index.call_args[0][0]
+
+    def test_if_the_annotation_isnt_in_the_DB_it_deletes_the_job_from_the_queue(
+        self,
+        annotations,
+        annotation_ids,
+        caplog,
+        db_session,
+        queue,
+    ):
+        for annotation in annotations[:limit]:
+            db_session.delete(annotation)
+
+        queue.add_all(
+            annotation_ids[:limit],
+            tag="test",
+            scheduled_at=five_minutes_ago,
+        )
+
+        queue.sync()
+
+        assert caplog.record_tuples == [
+            (
+                Any(),
+                Any(),
+                "Deleting 10 sync annotation jobs because their annotations have been deleted from the DB",
+            )
+        ]
```

Comment:
I think for monitoring and debugging purposes in production and dev I do want it to log _why_ it thinks it's deleting or indexing annotations, not just _that_ it is. And I think I do want tests for that logging -- I think the logging and its correctness is a significant part of the behaviour that I want from the method. Logging is going to be the main way that we test and monitor this job in production.

The tests do cover that the logging isn't lying. For example they set up a situation in which some jobs' annotations have been deleted from the DB. Then they test that the method both deleted the jobs from the queue _and_ logged that it was deleting them because their annotations were deleted from the DB, and didn't log anything else. So the test covers that the right log messages are logged at the right times. This also helps to make sure that the test set up is correct, and that the test hasn't accidentally set up some other situation that caused the method to delete the jobs for some other reason.

If you try breaking the logging you'll see that any breakage will cause a test to fail. E.g. change it to fail to log one of the log messages, or log the wrong message in the wrong circumstance, etc. The tests have it all covered.

I've changed the `caplog` fixture to only capture log messages from the module under test, so that if anything else that gets called (e.g. elasticsearch etc) logs anything that won't affect these tests.

I'm not concerned about changes to log messages causing tests to fail: it's unlikely that we'll want to change log messages in the future, and if we do the trivial test change will only take two seconds to make. Log message ordering shouldn't break any tests because each test only triggers one log message. It seems like the worst that can happen here is a trivial test breakage could cost a couple of seconds, and that seems unlikely to happen.

Reply:
There are a few different things mixed together here, both conceptually and in the tests.

### It's not clear in the code if we are testing the logs or using the logs to test

The first thing is it isn't clear if the assertions about the logs are intended to test the logging behavior, or intended to prove the behavior under test. Without you explaining this above it's not clear which intention you have. This is why I said the the logs can lie. Assuming that the code is wrong (which is what tests are about) we can't rely on the logging output to determine what it's doing, it's very circular.

It doesn't sound like that's what you were doing here, so that point is a bit moot, but the confusion is real.

### Fixing logging behaviour in tests increases costs, and makes adding/removing logging less desirable 

Here is a quick test. Add one line of logging to the top of the method and then run the tests. For me 9 of them fail. 

It's not at all clear why, and it causes the undesirable behaviour we were discussing with functional tests where you do one thing and a massive number of things blow up. (Suggestions on how to fix this below!) It also means that when you add a new line, presumably you're expected to add logging for it. It raises the costs all round.

This is because the tests don't just assert the content of the message  they also assert the position of the log message in the stream of logs. You can fix some of this by doing this instead:

``` 
        assert (
            "Syncing 1 annotations that are missing from Elasticsearch"
            in caplog.record_tuples[0]
        )
```

But this still leaves some lines asserting we don't log in certain situations. The point here is, we are locking down functionality that isn't is not, in the future worlds of what this code could be, (imho) important at all.

Is it mandatory that this code _never_ logs in certain situations, or is that just how it happens to behave right now? It is absolutely mandatory that it clears completed jobs etc., but I feel **these tests are enforcing non-functional requirements**.

### I'm not sold that this is how we are going to monitor this

I think this is how you're going to monitor it when you deploy it and look at it today and tomorrow. I think this is also useful debugging info going forward, but this isn't a monitoring solution.

If you want that we should be logging metrics somewhere. Looking for random things in Papertrail isn't how we do this. Which leads to...

## Assuming we're doing this, lets think about how to do it best

This log testing is currently testing two separate behaviours:

 * Collecting information about what situations we discovered to give insight about what we did and why
 * Testing a particular serialisation of that information

When you see assertions made about logging there are generally two questions to ask:

 * Are we testing behavior that doesn't matter?
 * If the behavior matters, how can I make this part of the system rather than a side effect?

Lets assume this is important and think about how we could structure this differently. I'd suggest:

 * Collect and return the stats in question
 * Test this in the individual methods
 * Have a separate test that checks this gets serialised as we'd expect

I've added an example of what that might look like here: https://github.com/hypothesis/h/pull/6309

With a setup like that we can check the counts returned in all of the separate unit tests, and then have one test like `def test_it_logs_the_counts()` which is the only one that would care about the exact serialisation.

This would also be a nice jumping off point if we do want to send these stats to NewRelic or some other collection service.

Reply:
* Using a counter does simplify the code quite a bit (only one call to `log.info()`) so I've done that: https://github.com/hypothesis/h/pull/6265/commits/9adf180b71806a16f642d679eef644354d10b5b8

* That commit also makes it so that changing the wording of the log messages will no longer break the tests

* I've also tweaked the assertions so that they don't break if more log lines are added to the method: https://github.com/hypothesis/h/pull/6265/commits/6df62d18704f278fceb9b7f6853aeda2adabc004

* I think it's clear enough that we're testing the logs rather than using the logs to test. Every time we make an assertion about something logged it's right next to an assertion about what actually happened, e.g.:

  ```python
  def test_if_the_annotation_is_missing_from_Elastic_it_indexes_it(self, ...):
      ...

      assert str({MISSING: 10}) in caplog.text
      batch_indexer.index.assert_called_once_with(Any.list.containing(annotation_ids).only())
  ```

* We seem to disagree on whether the logging in this particular function is an actual functional requirement or an incidental side-effect. I consider it to be a real functional requirement and I absolutely am intending to use this to monitor and debug this job in production. These log messages are going to be the only way that we can see what this function is up to -- how many jobs it's deleting and how many annotations it's syncing and why. That's going to be useful for initial testing of the job in production and over its first few weeks of deployment, and also for monitoring the job when it's recovering for us after an outage, or when a big reindex has been scheduled, etc. We could send metrics to New Relic and have a dashboard of graphs instead but I'm not planning to do that work--logging is the cheap-and-easy version. The only thing I'm planning to add to New Relic is basic queue size monitoring and a queue size alarm.

* I'd prefer to just use pytest's builtin `caplog` fixture for this. It's provided by pytest for this purpose, and it's pretty simple (`assert str({UP_TO_DATE: 10}) in caplog.text` etc). I think we should just use this and get on with it. It isn't really necessary here to separate information collected (the counts themselves) from serialization (the logged strings) and doing so would mean I'd have to write more tests and that the tests would be less direct, as there would have to be tests asserting about the returned counts and then separate tests asserting that the returned counts are also logged.

  Instead of using `caplog` we could patch `logger` and then the tests could do things like `logger.info.assert_called_with(expected_counts_dict)`. Rather than having the method return the counts dict for us to assert against. That way there's doesn't have to be a separate "and it also passes the counts that it returns to `logger.info()`" test. But honestly I think pylint has done this for us with `caplog`, and `assert str({UP_TO_DATE: 10}) in caplog.text` isn't much different from `logger.info.assert_called_with({UP_TO_DATE: 10})` (and it's actually more robust because it doesn't care what log level was used).